go/login/login.go

```diff
@@ -0,0 +1,388 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package login
+
+import (
+	"encoding/base64"
+	"fmt"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/google/glome/go/glome"
+)
+
+const (
+	// Minimal acceptable length of a handshake. 1 byte for the prefix, 32 bytes for the key
+	minHandshakeLen = 1 + glome.PublicKeySize
+)
+
+var (
+	// ErrInvalidURLFormat denotes that the URL has a wrong format.
+	ErrInvalidURLFormat = fmt.Errorf("URL is malformed")
+	// ErrInvalidHandshakeLen denotes that the handshake is too short.
+	ErrInvalidHandshakeLen = fmt.Errorf("handshake length is small: should be at least %d", minHandshakeLen)
+	// ErrVersionNotSupported denotes that the version of glome-login URL format is not supported.
+	ErrVersionNotSupported = fmt.Errorf("version not supported")
+	// ErrInvalidPrefixType denotes that the prefix type is invalid.
+	ErrInvalidPrefixType = fmt.Errorf("invalid prefix type")
+	// ErrIncorrectTag denotes that received tag is incorrect.
+	ErrIncorrectTag = fmt.Errorf("invalid tag")
+	// ErrResponseNotInitialized denotes that the response is not initialized.
+	ErrResponseNotInitialized = fmt.Errorf("response is not initialized")
+	// ErrServerKeyNotFound server key not found.
+	ErrServerKeyNotFound = fmt.Errorf("server key not found")
+)
+
+var (
+	validURLPrefix = regexp.MustCompile(`/(?P<v>v[1-9][0-9]*)/(?P<handshake>[\w=-]+)/`)
+)
+
+// Message struct represents the context required for authorization.
+// It contains HostIDType - type of identity, HostID - identity of the target (e.g. hostname, serial number, etc.),
+// Action - action that is being authorized.
+type Message struct {
+	HostIDType string
+	HostID     string
+	Action     string
+}
+
+// Construct returns a message from a Message according to the format: [<hostid-type>:]<hostid>[/<action>].
+// URL escaping is optional.
+func (m *Message) Construct(esc bool) []byte {
+	hostIDType := ""
+	hostID := ""
+	action := ""
+	if esc {
+		hostIDType = escape(m.HostIDType)
+		hostID = escape(m.HostID)
+	} else {
+		hostIDType = m.HostIDType
+		hostID = m.HostID
+	}
+
+	if hostIDType != "" {
+		hostIDType += ":"
+	}
+
+	if m.Action != "" {
+		action = "/" + m.Action
+	}
+	return []byte(hostIDType + hostID + action)
+}
+
+// Escapes the string so it can be safely placed inside a URL path segment,
+// replacing "/#?" special characters and not replacing "!*'();:@&=+$,[]" special characters.
```

Comment:
Filed https://github.com/google/glome/issues/62 so that we eventually address the imprecise wording in the documentation. I feel like we had more bits that needed fixing, but alas.